client/state/jetpack-connection-health/reducer.js

```diff
@@ -3,52 +3,93 @@ import {
 } from 'calypso/state/action-types';
+import {
+	combineReducers,
+	keyedReducer,
+	withPersistence,
+	withSchemaValidation,
+} from 'calypso/state/utils';
+import { schema } from './schema';
 
-export const jetpackConnectionHealth = ( state = {}, action ) => {
+/**
+ * `Reducer` function which handles Jetpack connection health state changes
+ * @param {Object} state - current state
+ * @param {Object} action - action
+ * @returns {{ is_healthy?: boolean, error?: string, jetpack_connection_problem?: boolean }} updated state - Jetpack connection health
+ */
+export const connectionHealth = withPersistence( ( state = {}, action ) => {
 	switch ( action.type ) {
 		case JETPACK_CONNECTION_HEALTHY: {
-			const { siteId } = action;
-
 			return {
 				...state,
-				[ siteId ]: {
-					...state[ siteId ],
-					jetpack_connection_problem: false,
-					is_healthy: true,
-					error: '',
-				},
+				jetpack_connection_problem: false,
+				error: '',
 			};
 		}
 
 		case JETPACK_CONNECTION_MAYBE_UNHEALTHY: {
-			const { siteId } = action;
-
 			return {
 				...state,
-				[ siteId ]: {
-					...state[ siteId ],
-					jetpack_connection_problem: true,
-				},
+				jetpack_connection_problem: true,
 			};
 		}
 
 		case JETPACK_CONNECTION_UNHEALTHY: {
-			const { siteId, errorCode } = action;
+			const { errorCode } = action;
 
 			return {
 				...state,
-				[ siteId ]: {
-					...state[ siteId ],
-					jetpack_connection_problem: true,
-					is_healthy: false,
-					error: errorCode,
-				},
+				jetpack_connection_problem: true,
+				error: errorCode,
 			};
 		}
 	}
 
 	return state;
-};
+} );
+
+/**
+ * `Reducer` function which handles request errors concerning Jetpack connection health API call.
+ * @param {string} state - current state
+ * @param {Object} action - action
+ * @returns {string} updated state - request error code
+ */
+export const requestError = withPersistence( ( state = '', action ) => {
+	switch ( action.type ) {
+		case JETPACK_CONNECTION_HEALTH_REQUEST:
+			return '';
+		case JETPACK_CONNECTION_HEALTH_REQUEST_FAILURE:
+			return action.error;
+		default:
+			return state;
+	}
+} );
+
+/**
+ * `Reducer` function which handles tracking the last request time for Jetpack connection health
+ * @param {number|null} state - current state
+ * @param {Object} action - action
+ * @returns {number|null} updated state - last connection health request time in milliseconds
+ */
+export const lastRequestTime = withPersistence( ( state = null, action ) => {
+	switch ( action.type ) {
+		case JETPACK_CONNECTION_HEALTH_REQUEST:
+			return action.lastRequestTime;
+		default:
+			return state;
+	}
+} );
+
+export const reducer = keyedReducer(
+	'siteId',
+	combineReducers( {
+		requestError,
+		connectionHealth,
+		lastRequestTime,
+	} )
+);
 
-export default withStorageKey( 'jetpackConnectionHealth', jetpackConnectionHealth );
+export default withStorageKey( 'jetpackConnectionHealth', withSchemaValidation( schema, reducer ) );
```

Comment:
The reducer structural changes will need some related unit test adjustments, particularly for the Jetpack Connect tests.

Reply:
@tyxla you mean the tests that were failing right? They are now fixed!